subprojects/plugins/src/main/java/org/gradle/api/internal/tasks/DefaultSourceSet.java

```diff
@@ -21,17 +21,23 @@
 import org.gradle.api.file.FileCollection;
 import org.gradle.api.file.FileTreeElement;
 import org.gradle.api.file.SourceDirectorySet;
+import org.gradle.api.internal.DynamicObjectAware;
+import org.gradle.api.internal.ExtensibleDynamicObject;
 import org.gradle.api.internal.jvm.ClassDirectoryBinaryNamingScheme;
 import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.plugins.Convention;
+import org.gradle.api.plugins.ExtensionContainer;
 import org.gradle.api.plugins.JavaPlugin;
 import org.gradle.api.specs.Spec;
 import org.gradle.api.tasks.SourceSet;
 import org.gradle.api.tasks.SourceSetOutput;
+import org.gradle.internal.metaobject.DynamicObject;
+import org.gradle.internal.reflect.Instantiator;
 import org.gradle.util.GUtil;
 
 import static org.gradle.util.ConfigureUtil.configure;
 
-public class DefaultSourceSet implements SourceSet {
+public class DefaultSourceSet implements SourceSet, DynamicObjectAware {
```

Comment:
❓ Why `DynamicObjectAware`?

Reply:
If you implement `ExtensionAware` you must also implement `DynamicObjectAware`.
I don't know the historical reasons as to why, but I know that the injector won't construct your object if you don't have this type.

@oehme You seem to understand much of the historical significance of many decisions, any insights on this one?

https://github.com/gradle/gradle/blob/e975872ae2cf1d521b73f1107cae772ca09c63cd/subprojects/core/src/main/java/org/gradle/api/internal/AbstractClassGenerator.java#L121-L126